File: CRISIS/CRISIS/Services/CrisisMonitorService.swift
import Foundation

protocol CrisisFeedService: Sendable {
    func fetchEvents() async throws -> [CrisisEvent]
}

struct CrisisMonitorService {
    private let feeds: [any CrisisFeedService]

    init(feeds: [any CrisisFeedService]? = nil, client: HTTPClienting = HTTPClient()) {
        if let feeds {
            self.feeds = feeds
        } else {
            self.feeds = [
                PoliticalFinancialNewsFeed(client: client, apiKey: AppConfig.newsAPIKey),
                GeopoliticalAlertCrisisFeed(client: client),
                FinancialStressCrisisFeed(client: client)
            ]
        }
    }

    func fetchEvents(limit: Int = 12) async throws -> [CrisisEvent] {
        var aggregated: [CrisisEvent] = []
        try await withThrowingTaskGroup(of: [CrisisEvent].self) { group in
            for feed in feeds {
                group.addTask {
                    try await feed.fetchEvents()
                }
            }

            for try await events in group {
                aggregated.append(contentsOf: events)
            }
        }

        let sorted = aggregated.sorted { $0.occurredAt > $1.occurredAt }
        return Array(sorted.prefix(limit))
    }
}

struct PoliticalFinancialNewsFeed: CrisisFeedService {
    private let client: HTTPClienting
    private let apiKey: String?
    private let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }()

    init(client: HTTPClienting = HTTPClient(), apiKey: String?) {
        self.client = client
        self.apiKey = apiKey
    }

    func fetchEvents() async throws -> [CrisisEvent] {
        guard let apiKey, apiKey.isEmpty == false else {
            return []
        }

        async let financialHeadlines = fetchTopHeadlines(category: "business", label: .financial, apiKey: apiKey)
        async let politicalHeadlines = fetchPoliticalNews(apiKey: apiKey)

        let (financial, political) = try await (financialHeadlines, politicalHeadlines)
        let merged = financial + political
        var unique: [CrisisEvent] = []
        var seenIDs = Set<String>()

        for event in merged {
            if seenIDs.insert(event.id).inserted {
                unique.append(event)
            }
        }

        // Preserve the order in which the different feeds were combined so that we keep
        // the expected "financial first, geopolitical second" layout that the UI (and tests)
        // rely on. The articles are already returned in descending order per feed.
        return Array(unique.prefix(10))
    }

    private func fetchTopHeadlines(category: String, label: CrisisCategory, apiKey: String) async throws -> [CrisisEvent] {
        var components = URLComponents()
        components.scheme = "https"
        components.host = "newsapi.org"
        components.path = "/v2/top-headlines"
        components.queryItems = [
            URLQueryItem(name: "language", value: "en"),
            URLQueryItem(name: "pageSize", value: "6"),
            URLQueryItem(name: "category", value: category)
        ]
        guard let request = makeRequest(from: components, apiKey: apiKey) else { return [] }
        let data = try await client.send(request)
        let response = try decoder.decode(NewsAPIResponse.self, from: data)
        return mapArticles(response.articles, category: label)
    }

    private func fetchPoliticalNews(apiKey: String) async throws -> [CrisisEvent] {
        var components = URLComponents()
        components.scheme = "https"
        components.host = "newsapi.org"
        components.path = "/v2/everything"
        components.queryItems = [
            URLQueryItem(name: "language", value: "en"),
            URLQueryItem(name: "pageSize", value: "6"),
            URLQueryItem(name: "sortBy", value: "publishedAt"),
            URLQueryItem(name: "q", value: "geopolitics OR government OR election OR crisis OR war")
        ]
        guard let request = makeRequest(from: components, apiKey: apiKey) else { return [] }
        let data = try await client.send(request)
        let response = try decoder.decode(NewsAPIResponse.self, from: data)
        return mapArticles(response.articles, category: .geopolitical)
    }

    private func makeRequest(from components: URLComponents, apiKey: String) -> URLRequest? {
        guard let url = components.url else { return nil }
        var request = URLRequest(url: url)
        request.addValue(apiKey, forHTTPHeaderField: "X-Api-Key")
        return request
    }

    private func mapArticles(_ articles: [NewsAPIArticle], category: CrisisCategory) -> [CrisisEvent] {
        articles.compactMap { article in
            guard let title = article.title else { return nil }
            let publishedAt = article.publishedAt ?? Date()
            let identifier = "\(category.rawValue)-\(title.hashValue)"
            return CrisisEvent(
                id: "news-\(identifier)",
                title: title,
                summary: article.description ?? article.content,
                region: article.source.name ?? "Weltweit",
                occurredAt: publishedAt,
                publishedAt: article.publishedAt,
                detailURL: nil,
                sourceName: article.source.name,
                source: .newsAPI,
                category: category,
                severityScore: severityScore(for: article.publishedAt)
            )
        }
    }

    private func severityScore(for date: Date?) -> Double {
        guard let date else { return 4 }
        let hours = Date().timeIntervalSince(date) / 3600
        switch hours {
        case ..<24: return 6
        case ..<72: return 5
        default: return 4
        }
    }
}

struct GeopoliticalAlertCrisisFeed: CrisisFeedService {
    private let client: HTTPClienting
    private let decoder = JSONDecoder()
    private let watchList: [(name: String, code: String)]

    init(
        client: HTTPClienting = HTTPClient(),
        watchList: [(String, String)] = [("Ukraine", "UKR"), ("Israel", "ISR"), ("Taiwan", "TWN"), ("South Africa", "ZAF"), ("Germany", "DEU")]
    ) {
        self.client = client
        self.watchList = watchList
    }

    func fetchEvents() async throws -> [CrisisEvent] {
        try await withThrowingTaskGroup(of: CrisisEvent?.self) { group in
            for entry in watchList {
                group.addTask {
                    try await self.fetchGovernanceAlert(for: entry)
                }
            }

            var events: [CrisisEvent] = []
            for try await result in group {
                if let event = result {
                    events.append(event)
                }
            }
            return events
        }
    }

    private func fetchGovernanceAlert(for entry: (name: String, code: String)) async throws -> CrisisEvent? {
        guard let url = URL(string: "https://api.worldbank.org/v2/country/\(entry.code)/indicator/PV.PSR.PIND?format=json&per_page=2") else {
            return nil
        }
        guard let data = try? await client.get(url),
              let response = try? decoder.decode(WorldBankValueResponse.self, from: data),
              let latest = response.entries.first(where: { $0.value != nil }),
              let value = latest.value,
              value < CrisisThresholds.politicalInstabilityCutoff else {
            return nil
        }

        let yearDate = makeYearDate(from: latest.date)

        return CrisisEvent(
            id: "geo-\(entry.code)",
            title: "Politische Instabilität \(entry.name)",
            summary: "Governance-Index \(value.formatted(.number.precision(.fractionLength(2))))",
            region: entry.name,
            occurredAt: yearDate,
            publishedAt: yearDate,
            detailURL: URL(string: "https://data.worldbank.org/indicator/PV.PSR.PIND"),
            sourceName: "World Bank",
            source: .worldBankGovernance,
            category: .geopolitical,
            severityScore: abs(value) * 2
        )
    }
}

struct FinancialStressCrisisFeed: CrisisFeedService {
    private let client: HTTPClienting
    private let decoder = JSONDecoder()
    private let watchList: [(name: String, code: String)]

    init(
        client: HTTPClienting = HTTPClient(),
        watchList: [(String, String)] = [("Germany", "DEU"), ("United States", "USA"), ("United Kingdom", "GBR"), ("Japan", "JPN"), ("China", "CHN")]
    ) {
        self.client = client
        self.watchList = watchList
    }

    func fetchEvents() async throws -> [CrisisEvent] {
        try await withThrowingTaskGroup(of: CrisisEvent?.self) { group in
            for entry in watchList {
                group.addTask {
                    try await self.fetchAlert(for: entry)
                }
            }

            var events: [CrisisEvent] = []
            for try await result in group {
                if let event = result {
                    events.append(event)
                }
            }
            return events
        }
    }

    private func fetchAlert(for entry: (name: String, code: String)) async throws -> CrisisEvent? {
        guard let url = URL(string: "https://api.worldbank.org/v2/country/\(entry.code)/indicator/NY.GDP.MKTP.KD.ZG?format=json&per_page=2") else {
            return nil
        }
        guard let data = try? await client.get(url),
              let response = try? decoder.decode(WorldBankValueResponse.self, from: data),
              let latest = response.entries.first(where: { $0.value != nil }),
              let value = latest.value,
              value < CrisisThresholds.recessionGrowthCutoff else {
            return nil
        }

        let yearDate = makeYearDate(from: latest.date)

        return CrisisEvent(
            id: "finance-\(entry.code)",
            title: "Rezession \(entry.name)",
            summary: "Reales BIP-Wachstum \(value.formatted(.number.precision(.fractionLength(1))))%",
            region: entry.name,
            occurredAt: yearDate,
            publishedAt: yearDate,
            detailURL: URL(string: "https://data.worldbank.org/indicator/NY.GDP.MKTP.KD.ZG"),
            sourceName: "World Bank",
            source: .worldbankFinance,
            category: .financial,
            severityScore: abs(value)
        )
    }
}

private struct NewsAPIResponse: Decodable {
    let articles: [NewsAPIArticle]
}

private struct NewsAPIArticle: Decodable {
    struct Source: Decodable {
        let name: String?
    }

    let title: String?
    let description: String?
    let content: String?
    let source: Source
    let publishedAt: Date?
    let url: URL?
}

private struct WorldBankValueResponse: Decodable {
    struct Entry: Decodable {
        let date: String
        let value: Double?
    }

    let entries: [Entry]

    init(from decoder: Decoder) throws {
        var container = try decoder.unkeyedContainer()
        _ = try? container.decode(DecodableIgnored.self)
        entries = try container.decode([Entry].self)
    }
}

private struct DecodableIgnored: Decodable { }

private extension String {
    var severityScore: Int {
        switch lowercased() {
        case "extreme": return 6
        case "severe": return 5
        case "moderate": return 4
        case "minor": return 3
        default: return 2
        }
    }
}

private func makeYearDate(from year: String) -> Date {
    Calendar.current.date(from: DateComponents(year: Int(year) ?? 2024)) ?? Date()
}

File: android/core-data/src/main/java/de/vibecode/crisis/core/data/CrisisMonitorService.kt
package de.vibecode.crisis.core.data

import de.vibecode.crisis.core.domain.CrisisSeverity
import de.vibecode.crisis.core.model.CrisisCategory
import de.vibecode.crisis.core.model.CrisisEvent
import de.vibecode.crisis.core.model.CrisisSettings
import de.vibecode.crisis.core.model.CrisisWatchlists
import de.vibecode.crisis.core.model.DataSource
import de.vibecode.crisis.core.model.WatchlistCountry
import de.vibecode.crisis.core.network.NewsApi
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.datetime.Clock
import kotlinx.datetime.Instant
import kotlinx.datetime.LocalDate
import kotlinx.datetime.TimeZone
import kotlinx.datetime.atStartOfDayIn
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.Request

interface CrisisFeedService {
    suspend fun fetchEvents(settings: CrisisSettings): List<CrisisEvent>
}

class CrisisMonitorService(private val feeds: List<CrisisFeedService>) {
    suspend fun fetchEvents(settings: CrisisSettings, limit: Int = 12): List<CrisisEvent> = coroutineScope {
        val results = feeds.map { feed ->
            async { feed.fetchEvents(settings) }
        }.flatMap { it.await() }

        results.sortedByDescending { it.occurredAt }.take(limit)
    }
}

class PoliticalFinancialNewsFeed(
    private val api: NewsApi,
    private val apiKey: String?,
    private val cache: NewsCache?
) : CrisisFeedService {
    override suspend fun fetchEvents(settings: CrisisSettings): List<CrisisEvent> {
        if (apiKey.isNullOrBlank()) {
            return cache?.load()?.events ?: emptyList()
        }

        return runCatching {
            val key = apiKey!!
            val financial = api.topHeadlines(category = "business", apiKey = key)
            val political = api.everything(query = AppConfig.NEWS_QUERY, apiKey = key)

            val merged = financial.articles.mapNotNull { article ->
                mapArticle(article, CrisisCategory.FINANCIAL)
            } + political.articles.mapNotNull { article ->
                mapArticle(article, CrisisCategory.GEOPOLITICAL)
            }

            val unique = LinkedHashMap<String, CrisisEvent>()
            merged.forEach { event ->
                if (!unique.containsKey(event.id)) {
                    unique[event.id] = event
                }
            }

            val events = unique.values.take(10)
            cache?.persist(events)
            events
        }.getOrElse {
            cache?.load()?.events ?: emptyList()
        }
    }

    private fun mapArticle(article: de.vibecode.crisis.core.network.NewsApiArticle, category: CrisisCategory): CrisisEvent? {
        val title = article.title ?: return null
        val publishedAt = article.publishedAt?.let { runCatching { Instant.parse(it) }.getOrNull() } ?: Clock.System.now()
        val identifier = "$category-${title.hashCode()}"
        return CrisisEvent(
            id = "news-$identifier",
            title = title,
            summary = article.description ?: article.content,
            region = article.source.name ?: "Weltweit",
            occurredAt = publishedAt,
            publishedAt = publishedAt,
            detailUrl = article.url,
            sourceName = article.source.name,
            source = DataSource.NEWS_API,
            category = category,
            severityScore = CrisisSeverity.newsSeverityScore(publishedAt)
        )
    }
}

class GeopoliticalAlertCrisisFeed(
    private val client: OkHttpClient,
    private val json: Json
) : CrisisFeedService {
    override suspend fun fetchEvents(settings: CrisisSettings): List<CrisisEvent> = coroutineScope {
        val active = filteredWatchlist(CrisisWatchlists.geopolitical, settings.geopoliticalWatchlist)
        if (active.isEmpty()) return@coroutineScope emptyList()
        active.map { entry ->
            async { fetchGovernanceAlert(entry, settings) }
        }.mapNotNull { it.await() }
    }

    private suspend fun fetchGovernanceAlert(entry: WatchlistCountry, settings: CrisisSettings): CrisisEvent? {
        val name = entry.name
        val code = entry.code
        val url = "https://api.worldbank.org/v2/country/$code/indicator/PV.PSR.PIND?format=json&per_page=2"
        val latest = fetchWorldBankValue(url) ?: return null
        val value = latest.value
        if (value >= settings.thresholdProfile.governanceCutoff) return null

        val yearDate = yearInstant(latest.year)
        return CrisisEvent(
            id = "geo-$code",
            title = "Politische Instabilitaet $name",
            summary = "Governance-Index ${"%.2f".format(value)}",
            region = name,
            occurredAt = yearDate,
            publishedAt = yearDate,
            detailUrl = "https://data.worldbank.org/indicator/PV.PSR.PIND",
            sourceName = "World Bank",
            source = DataSource.WORLD_BANK_GOVERNANCE,
            category = CrisisCategory.GEOPOLITICAL,
            severityScore = CrisisSeverity.governanceSeverity(value)
        )
    }

    private suspend fun fetchWorldBankValue(url: String): WorldBankValue? {
        val body = getBody(url)
        if (body.isBlank()) return null
        val root = json.parseToJsonElement(body).jsonArray
        val entries = root.getOrNull(1)?.jsonArray ?: return null
        val latest = entries.firstOrNull { it.jsonObject["value"]?.jsonPrimitive?.doubleOrNull != null } ?: return null
        val year = latest.jsonObject["date"]?.jsonPrimitive?.content?.toIntOrNull() ?: return null
        val value = latest.jsonObject["value"]?.jsonPrimitive?.doubleOrNull ?: return null
        return WorldBankValue(year, value)
    }

    private suspend fun getBody(url: String): String = kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
        val request = Request.Builder().url(url).build()
        client.newCall(request).execute().use { response ->
            response.body?.string().orEmpty()
        }
    }
}

class FinancialStressCrisisFeed(
    private val client: OkHttpClient,
    private val json: Json
) : CrisisFeedService {
    override suspend fun fetchEvents(settings: CrisisSettings): List<CrisisEvent> = coroutineScope {
        val active = filteredWatchlist(CrisisWatchlists.financial, settings.financialWatchlist)
        if (active.isEmpty()) return@coroutineScope emptyList()
        active.map { entry ->
            async { fetchAlert(entry, settings) }
        }.mapNotNull { it.await() }
    }

    private suspend fun fetchAlert(entry: WatchlistCountry, settings: CrisisSettings): CrisisEvent? {
        val name = entry.name
        val code = entry.code
        val url = "https://api.worldbank.org/v2/country/$code/indicator/NY.GDP.MKTP.KD.ZG?format=json&per_page=2"
        val latest = fetchWorldBankValue(url) ?: return null
        val value = latest.value
        if (value >= settings.thresholdProfile.recessionCutoff) return null

        val yearDate = yearInstant(latest.year)
        return CrisisEvent(
            id = "finance-$code",
            title = "Rezession $name",
            summary = "Reales BIP-Wachstum ${"%.1f".format(value)}%",
            region = name,
            occurredAt = yearDate,
            publishedAt = yearDate,
            detailUrl = "https://data.worldbank.org/indicator/NY.GDP.MKTP.KD.ZG",
            sourceName = "World Bank",
            source = DataSource.WORLD_BANK_FINANCE,
            category = CrisisCategory.FINANCIAL,
            severityScore = CrisisSeverity.recessionSeverity(value)
        )
    }

    private suspend fun fetchWorldBankValue(url: String): WorldBankValue? {
        val body = getBody(url)
        if (body.isBlank()) return null
        val root = json.parseToJsonElement(body).jsonArray
        val entries = root.getOrNull(1)?.jsonArray ?: return null
        val latest = entries.firstOrNull { it.jsonObject["value"]?.jsonPrimitive?.doubleOrNull != null } ?: return null
        val year = latest.jsonObject["date"]?.jsonPrimitive?.content?.toIntOrNull() ?: return null
        val value = latest.jsonObject["value"]?.jsonPrimitive?.doubleOrNull ?: return null
        return WorldBankValue(year, value)
    }

    private suspend fun getBody(url: String): String = kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
        val request = Request.Builder().url(url).build()
        client.newCall(request).execute().use { response ->
            response.body?.string().orEmpty()
        }
    }
}

private data class WorldBankValue(val year: Int, val value: Double)

private fun filteredWatchlist(
    candidates: List<WatchlistCountry>,
    enabledCodes: Set<String>
): List<WatchlistCountry> {
    if (enabledCodes.isEmpty()) return emptyList()
    return candidates.filter { enabledCodes.contains(it.code) }
}

private fun yearInstant(year: Int): Instant {
    val date = LocalDate(year, 1, 1)
    return date.atStartOfDayIn(TimeZone.UTC)
}

